sensor: Add ICM42686 Driver Support

ubieda opened the pull request:
### Description
This PR adds support for ICM42686 driver, by extending and refactoring existing ICM42688 driver.
The new variant contains feature parity with the original driver:
- Fetch and Get (Sensor Model V1)
- Sensor Model V2:
    - Read/Decode
    - Streaming mode
- Build Time Configuration thorugh device-tree properties
- Run Time Configuration through attribute set/get API.

### Testing
This was tested on VMU RT1170, which contains both ICM42688 and ICM42686, running the sensor-shell sample to set up and obtain sensor readings:

- Build command:
```
west build -b vmu_rt1170/mimxrt1176/cm7 samples/sensor/sensor_shell
```
- Sensor Read/Decode:
```
uart:~$ sensor get icm42688p0@0 
channel type=0(accel_x) index=0 shift=8 num_samples=1 value=19162868474ns (-0.004787)
channel type=1(accel_y) index=0 shift=8 num_samples=1 value=19162868474ns (0.052671)
channel type=2(accel_z) index=0 shift=8 num_samples=1 value=19162868474ns (9.916782)
channel type=3(accel_xyz) index=0 shift=8 num_samples=1 value=19162868474ns, (-0.004787, 0.052671, 9.916782)
channel type=4(gyro_x) index=0 shift=6 num_samples=1 value=19162868474ns (-0.004255)
channel type=5(gyro_y) index=0 shift=6 num_samples=1 value=19162868474ns (-0.026604)
channel type=6(gyro_z) index=0 shift=6 num_samples=1 value=19162868474ns (0.012769)
channel type=7(gyro_xyz) index=0 shift=6 num_samples=1 value=19162868474ns, (-0.004255, -0.026604, 0.012769)
channel type=12(die_temp) index=0 shift=9 num_samples=1 value=19162868474ns (31.951991)
uart:~$ sensor get icm
  icm42688p0@0  icm42686p0@0  
uart:~$ sensor get icm42686p0@0 
channel type=0(accel_x) index=0 shift=8 num_samples=1 value=3300985903ns (0.052671)
channel type=1(accel_y) index=0 shift=8 num_samples=1 value=3300985903ns (0.076613)
channel type=2(accel_z) index=0 shift=8 num_samples=1 value=3300985903ns (-9.720457)
channel type=3(accel_xyz) index=0 shift=8 num_samples=1 value=3300985903ns, (0.052671, 0.076613, -9.720457)
channel type=4(gyro_x) index=0 shift=6 num_samples=1 value=3300985903ns (0.000000)
channel type=5(gyro_y) index=0 shift=6 num_samples=1 value=3300985903ns (-0.006384)
channel type=6(gyro_z) index=0 shift=6 num_samples=1 value=3300985903ns (0.004255)
channel type=7(gyro_xyz) index=0 shift=6 num_samples=1 value=3300985903ns, (0.000000, -0.006384, 0.004255)
channel type=12(die_temp) index=0 shift=9 num_samples=1 value=3300985903ns (35.084540)
```
- Streaming Mode:
```
uart:~$ sensor attr_set icm42686p0@0 accel_xyz sampling_frequency 25
icm42686p0@0 channel=accel_xyz, attr=sampling_frequency set to value=25
[00:03:17.412,000] <inf> ICM4268X_LL: FIFO ENABLED
uart:~$ sensor attr_set icm42686p0@0 gyro_xyz sampling_frequency 25
icm42686p0@0 channel=gyro_xyz, attr=sampling_frequency set to value=25
[00:03:19.423,000] <inf> ICM4268X_LL: FIFO ENABLED
uart:~$ sensor attr_set icm42686p0@0 all batch_dur 10000
icm42686p0@0 channel=all, attr=batch_dur set to value=10000
[00:03:21.782,000] <inf> ICM4268X_LL: FIFO ENABLED
uart:~$ sensor attr_set icm42686p0@0 all batch_dur 10000
uart:~$ sensor stream icm
  icm42688p0@0  icm42686p0@0
uart:~$ sensor stream icm42686p0@0 on fifo_wm incl
Disabling existing stream
Enabling stream...
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
Trigger (11 / fifo_full) detected
channel type=0(accel_x) index=0 shift=9 num_samples=100 value=2940457166ns (0.081131)
channel type=1(accel_y) index=0 shift=9 num_samples=100 value=2940457166ns (0.081131)
channel type=2(accel_z) index=0 shift=9 num_samples=100 value=2940457166ns (0.081131)
channel type=3(accel_xyz) index=0 shift=9 num_samples=100 value=2940457166ns, (0.081131, 0.046600, -9.702820)
channel type=4(gyro_x) index=0 shift=7 num_samples=100 value=2940457166ns (-0.005792)
channel type=5(gyro_y) index=0 shift=7 num_samples=100 value=2940457166ns (-0.005792)
channel type=6(gyro_z) index=0 shift=7 num_samples=100 value=2940457166ns (-0.005792)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=100 value=2940457166ns, (-0.005792, 0.000636, 0.004793)
channel type=12(die_temp) index=0 shift=9 num_samples=100 value=2940457166ns (35.059555)
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
channel type=0(accel_x) index=0 shift=9 num_samples=25 value=2430948229ns (0.082568)
channel type=1(accel_y) index=0 shift=9 num_samples=25 value=2430948229ns (0.082568)
channel type=2(accel_z) index=0 shift=9 num_samples=25 value=2430948229ns (0.082568)
channel type=3(accel_xyz) index=0 shift=9 num_samples=25 value=2430948229ns, (0.082568, 0.047605, -9.702461)
channel type=4(gyro_x) index=0 shift=7 num_samples=25 value=2430948229ns (-0.005712)
channel type=5(gyro_y) index=0 shift=7 num_samples=25 value=2430948229ns (-0.005712)
channel type=6(gyro_z) index=0 shift=7 num_samples=25 value=2430948229ns (-0.005712)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=25 value=2430948229ns, (-0.005712, 0.000575, 0.004859)
channel type=12(die_temp) index=0 shift=9 num_samples=25 value=2430948229ns (35.067028)
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
channel type=0(accel_x) index=0 shift=9 num_samples=25 value=3421436652ns (0.083047)
channel type=1(accel_y) index=0 shift=9 num_samples=25 value=3421436652ns (0.083047)
channel type=2(accel_z) index=0 shift=9 num_samples=25 value=3421436652ns (0.083047)
channel type=3(accel_xyz) index=0 shift=9 num_samples=25 value=3421436652ns, (0.083047, 0.045403, -9.703323)
channel type=4(gyro_x) index=0 shift=7 num_samples=25 value=3421436652ns (-0.005541)
channel type=5(gyro_y) index=0 shift=7 num_samples=25 value=3421436652ns (-0.005541)
channel type=6(gyro_z) index=0 shift=7 num_samples=25 value=3421436652ns (-0.005541)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=25 value=3421436652ns, (-0.005541, 0.000500, 0.004923)
channel type=12(die_temp) index=0 shift=9 num_samples=25 value=3421436652ns (35.065820)
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
channel type=0(accel_x) index=0 shift=9 num_samples=25 value=4411926607ns (0.079407)
channel type=1(accel_y) index=0 shift=9 num_samples=25 value=4411926607ns (0.079407)
channel type=2(accel_z) index=0 shift=9 num_samples=25 value=4411926607ns (0.079407)
channel type=3(accel_xyz) index=0 shift=9 num_samples=25 value=4411926607ns, (0.079407, 0.047989, -9.701982)
channel type=4(gyro_x) index=0 shift=7 num_samples=25 value=4411926607ns (-0.005882)
channel type=5(gyro_y) index=0 shift=7 num_samples=25 value=4411926607ns (-0.005882)
channel type=6(gyro_z) index=0 shift=7 num_samples=25 value=4411926607ns (-0.005882)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=25 value=4411926607ns, (-0.005882, 0.000564, 0.004699)
channel type=12(die_temp) index=0 shift=9 num_samples=25 value=4411926607ns (35.072161)
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
channel type=0(accel_x) index=0 shift=9 num_samples=25 value=5402416407ns (0.080269)
channel type=1(accel_y) index=0 shift=9 num_samples=25 value=5402416407ns (0.080269)
channel type=2(accel_z) index=0 shift=9 num_samples=25 value=5402416407ns (0.080269)
channel type=3(accel_xyz) index=0 shift=9 num_samples=25 value=5402416407ns, (0.080269, 0.046935, -9.703132)
channel type=4(gyro_x) index=0 shift=7 num_samples=25 value=5402416407ns (-0.005978)
channel type=5(gyro_y) index=0 shift=7 num_samples=25 value=5402416407ns (-0.005978)
channel type=6(gyro_z) index=0 shift=7 num_samples=25 value=5402416407ns (-0.005978)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=25 value=5402416407ns, (-0.005978, 0.000660, 0.004849)
channel type=12(die_temp) index=0 shift=9 num_samples=25 value=5402416407ns (35.067934)
Trigger (1 / data_ready) detected
Trigger (10 / fifo_wm) detected
channel type=0(accel_x) index=0 shift=9 num_samples=25 value=6392905612ns (0.079886)
channel type=1(accel_y) index=0 shift=9 num_samples=25 value=6392905612ns (0.079886)
channel type=2(accel_z) index=0 shift=9 num_samples=25 value=6392905612ns (0.079886)
channel type=3(accel_xyz) index=0 shift=9 num_samples=25 value=6392905612ns, (0.079886, 0.044349, -9.701791)
channel type=4(gyro_x) index=0 shift=7 num_samples=25 value=6392905612ns (-0.005850)
channel type=5(gyro_y) index=0 shift=7 num_samples=25 value=6392905612ns (-0.005850)
channel type=6(gyro_z) index=0 shift=7 num_samples=25 value=6392905612ns (-0.005850)
channel type=7(gyro_xyz) index=0 shift=7 num_samples=25 value=6392905612ns, (-0.005850, 0.000532, 0.004955)
channel type=12(die_temp) index=0 shift=9 num_samples=25 value=6392905612ns (35.068537)
uart:~$ sensor stream icm42686p0@0 off
Disabling existing stream
uart:~$ 
```

ubieda committed:
general: icm4268x: Refactor ICM42688 driver to ICM4268X

As a first step to enable the similar variants (e.g: ICM42686),
refactor common functionality into icm4268x files. As a result,
applications using the icm42688 will need to have both compatible
properties: "invensense,icm42688" and "invensense,icm4268x" defined.
In-tree boards have been modified to comply with this pattern.

This patch does not contain functional changes. The driver should
work the same as before.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: icm4268x: Add support for ICM42686 variant

Now this driver supports both ICM42688 and ICM42686.
Tested with read-decode as well as streaming mode.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
boards: vmu_rt1170: Update dts to reflect the second IMU being ICM42686

Additionally, remove overlay from sensor-shell sample, which is
redundant.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: sensor: Add icm42686 node to test

For build-time validation.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
sensor: icm4568x: Address existing code issues by static analysis

Addressing low-hanging fruits.

Put in a separate commit in order to make it easier to keep track of
changes.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

bperseghetti reviewed (dismissed):
Great addition! I also tested this on real hardware, both targets (42688 and 42686) work perfectly.

ubieda committed:
doc: migration guide 4.3: Add entry to continue using icm42688

Users now requiring to define both dt compatibles: both icm4268x and
icm42688 to work.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

teburd commented on drivers/sensor/tdk/icm4268x/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+zephyr_library_sources_ifdef(CONFIG_SENSOR_ASYNC_API icm4268x_rtio.c)
+zephyr_library_sources_ifdef(CONFIG_ICM4268X_DECODER icm4268x_decoder.c)
+zephyr_library_sources_ifdef(CONFIG_ICM4268X_STREAM icm4268x_rtio_stream.c)
+zephyr_library_sources_ifdef(CONFIG_ICM4268X_TRIGGER icm4268x_trigger.c)
```
Of all the APIs the Sensor API has currently, I believe the trigger API is the least usable among them and conflicts the most with the stream API. I'm wondering if perhaps this API is one we really should aim to deprecate.

Supporting the polling read and stream is fairly easy I think. Supporting trigger and stream seems conflict ridden.

kartben commented on drivers/sensor/tdk/icm4268x/icm4268x_decoder.c:
```diff
@@ -735,13 +751,13 @@ static bool icm24688_decoder_has_trigger(const uint8_t *buffer, enum sensor_trig
+	.get_frame_count = icm4268x_decoder_get_frame_count,
+	.get_size_info = icm4268x_decoder_get_size_info,
+	.decode = icm4268x_decoder_decode,
 	.has_trigger = icm24688_decoder_has_trigger,
```
would be nice to get that typo fixed

ubieda replied:
Thanks for spotting this. Addressed in #93742

kartben commented on drivers/sensor/tdk/icm4268x/icm4268x.h:
```diff
@@ -0,0 +1,582 @@
+		out->val2 = 0;
+		return;
+	case ICM4268X_DT_ACCEL_ODR_16000:
+		out->val1 = 1600;
```
```suggestion
		out->val1 = 16000;
```

ubieda replied:
Thanks for spotting this. Addressed in #93742
